flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1697,133 +1625,486 @@ class ExprLowering {
+    return std::visit(
+        [&](const auto &e) {
+          auto loc = getLoc();
+          auto f = genarr(e);
+          auto [iterSpace, insPt] = genIterSpace();
+          auto exv = f(iterSpace);
+          auto innerArg = iterSpace.innerArgument();
+          auto upd = builder.create<fir::ArrayUpdateOp>(
+              loc, innerArg.getType(), innerArg, fir::getBase(exv),
+              iterSpace.iterVec());
+          builder.create<fir::ResultOp>(loc, upd.getResult());
+          builder.restoreInsertionPoint(insPt);
+          return iterSpace.outerResult();
+        },
+        exp.u);
+  }
+
+  std::pair<IterationSpace, mlir::OpBuilder::InsertPoint> genIterSpace() {
+    auto loc = getLoc();
+    auto idxTy = builder.getIndexType();
+    auto zero = builder.createIntegerConstant(loc, idxTy, 0);
+    auto one = builder.createIntegerConstant(loc, idxTy, 1);
+    llvm::SmallVector<mlir::Value, 8> loopUppers;
+
+    // Convert the shape to closed interval form.
+    for (auto s : destination.getExtents()) {
+      auto ub = builder.createConvert(loc, idxTy, s);
+      auto up = builder.create<mlir::SubIOp>(loc, ub, one);
+      loopUppers.push_back(up);
+    }
+
+    // Iteration space is created with outermost columns, innermost rows
+    fir::DoLoopOp inner;
+    auto innerArg = destination.getResult();
+    mlir::Value outerRes;
+    llvm::SmallVector<mlir::Value, 8> ivars;
+    auto insPt = builder.saveInsertionPoint();
+    const auto loopDepth = loopUppers.size();
+    assert(loopDepth > 0);
+    for (auto i : llvm::enumerate(llvm::reverse(loopUppers))) {
+      if (i.index() > 0) {
+        assert(inner);
+        builder.setInsertionPointToStart(inner.getBody());
+      }
+      auto loop = builder.create<fir::DoLoopOp>(
+          loc, zero, i.value(), one, /*unordered=*/false,
+          /*finalCount=*/false, mlir::ValueRange{innerArg});
+      innerArg = loop.getRegionIterArgs().front();
+      ivars.push_back(loop.getInductionVar());
+      if (!outerRes)
+        outerRes = loop.getResult(0);
+      if (!inner)
+        insPt = builder.saveInsertionPoint();
+      if (i.index() < loopDepth - 1) {
+        // Add the fir.result for all loops except the innermost one.
+        builder.setInsertionPointToStart(loop.getBody());
+        builder.create<fir::ResultOp>(loc, innerArg);
+      }
+      inner = loop;
+    }
+
+    // move insertion point inside loop nest
+    builder.setInsertionPointToStart(inner.getBody());
+    return {IterationSpace{innerArg, outerRes, ivars}, insPt};
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Expression traversal and lowering.
+  //===--------------------------------------------------------------------===//
+
+  // Lower the expression in a scalar context.
+  template <typename A>
+  ExtValue asScalar(const A &x) {
+    return ExprLowering{getLoc(), converter, symMap, stmtCtx}.genval(x);
```

Comment:
I am not following by which magic this is computing the scalar outside of the loop (obviously, it is computed outside of the loops given your tests, so it works :) )

Reply:
Ah, found it ! it is not being generated in a lamda :)